lib/galaxy/managers/users.py

```diff
@@ -840,13 +837,20 @@ def _add_parsers(self):
         self.fn_filter_parsers.update({})
 
 
-def get_user_by_username(session, user_class, username):
-    """
-    Get a user from the database by username.
-    (We pass the session and the user_class to accommodate usage from the tool_shed app.)
-    """
-    try:
-        stmt = select(user_class).filter(user_class.username == username)
-        return session.execute(stmt).scalar_one()
-    except Exception:
-        return None
+def get_users_by_ids(session: Session, user_ids):
+    stmt = select(User).where(User.id.in_(user_ids))
+    return session.scalars(stmt).all()
+
+
+# The get_user_by_email and get_user_by_username functions may be called from
+# the tool_shed app, which has its own User model, which is different from
+# galaxy.model.User. In that case, the tool_shed user model should be passed as
+# the model_class argument.
+def get_user_by_email(session, email: str, model_class=User):
+    stmt = select(model_class).filter(model_class.email == email).limit(1)
+    return session.scalars(stmt).first()
```

Comment:
Is that the same as `get_user_by_username` ? It effectively looks like `.one_or_none()` ?

Reply:
`one_or_none` will raise an exception on multiple results, `first` will not. The `username` attribute is unique, but `email` is not; using `one_or_none` would thus raise an error in a potentially legitimate case, I think. I didn't want to change this behavior without looking closer at all the use cases within this PR. If you'd rather I look into this within this PR, I can do that, of course.

Reply:
sorry, meant to write this on `get_user_by_username`. previously no result returned None, now it will blow up, so that behavior has changed.

Reply:
Sorry, why will it blow up? `first` returns first result of `None` if no result is present. [Same behavior as before](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/users.py#L843). What am I missing?

Reply:
Ooops, didn't no that first behaves like one_or_none